### DISCUSSION TITLE
Where are the communities for disconnected nodes of a graph?

### DISCUSSION THREAD

**Discussion by JohannHM** [mapequation/infomap#194] [Q&A] [answered]
Thank so much for this product. As novice user of communities a simple question comes to my mind.
What happen with disconnected nodes of a network when computing infomap communities?
The algorithm runs fine when works with a giant component, but when I work with a disconnected graph, that includes disconnected and isolated nodes, they suddenly disappear in the infomap output. It looks like infomap internally remove them out in some way before runs the infomap algorithm. I was thinking that disconnected nodes receive a proper communities...
Are they (disconnected nodes) always removed when performing infomap?

Thanks in advance for your answers!

**Comment by danieledler** [comment]
Hi @JohannHM and thanks for your question! Disconnected nodes usually end up in singleton modules (one module for each disconnected node). They should not be removed by Infomap at any stage, so can you please provide an example of what you get?

**Comment by JohannHM** [comment, marked as answer]
Hi @danieledler thank for your opinion.
Im working with cdlib library that uses exactly the same code here. In fact this python example was tested on your web of infomap with the same results. The idea is attacking a network by removing some hubs, thus some disconnected nodes appear, and then applying infomap. Recovering the nodes after infomap with those before applying it. Look at this:

```python
import networkx as nx
import math

g = nx.karate_club_graph()                  # karate-club network  
k1 = dict(g.degree())                       # taking its degree k1

percentage = 10                                 # percentage of nodes to remove
n_nodes = math.floor(percentage*len(k1)/100)    # num of nodes to remove
nodes_2_remove = list(k1.keys())[:n_nodes]      # nodes to remove

G = g.copy()                                # taking a copy of graph to attack it
G.remove_nodes_from(nodes_2_remove)         # attaking the graph
k2 = dict(G.degree())                       # taking degree k2 of attaked graph

nodes_degree_zero = [n for n, k in k2.items() if k == 0] # look for nodes with degree = 0

from cdlib import algorithms                # importing the infomap
p = algorithms.infomap(G).communities       # applying infomap

recovering_nodes = []                       # recovering nodes from infomap communities
for community in p:
    recovering_nodes.extend(community)
recovering_nodes = sorted(recovering_nodes)

# comparing nodes before applying infomap with those after infomap method
missing_nodes = sorted(list(set(G.nodes()) - set(recovering_nodes)))

# infomap does not take into disconnected ndoes as singleton communities
print(missing_nodes == nodes_degree_zero)
```

when I testes with the web https://www.mapequation.org/infomap/
Same results.

thanks

**Reply by danieledler** [reply]
I see, thanks for more details! You are using infomap through a third-party library `cdlib` that unfortunately [only loop over the edges](https://github.com/GiulioRossetti/cdlib/blob/master/cdlib/algorithms/crisp_partition.py#L720) in the graph when they add the network to Infomap. This is why disconnected nodes are missing in the output.

I didn't know about that library before, thanks for the information, I will make a pull request so they can fix that, and also suggest that they allow parameters to be sent to Infomap to support for example directed networks.

Until then, you can use Infomap directly, with the Python API documented [here](https://mapequation.github.io/infomap/python/). To include disconnected nodes you just have to call [add_node](https://mapequation.github.io/infomap/python/infomap.html#infomap.Infomap.add_node).

On the Infomap web app, disconnected nodes should be included in the output. How did you test it when you got the same result?

**Reply by antoneri** [reply]
I was curious, so I tried this: a network with no links to node 5

```python
*Vertices 6
1 "1"
2 "2"
3 "3"
4 "4"
5 "5"
6 "6"
*Edges 10
1 2
1 3
1 4
2 1
2 3
3 2
3 1
4 1
4 6
6 4
```

I get the following:

```python
# v1.3.0
# ./Infomap network . --clu --ftree
# started at 2021-05-28 14:36:56
# completed in 0.107 s
# partitioned into 2 levels with 2 top modules
# codelength 2.24644 bits
# relative codelength savings 0%
# module level 1
# node_id module flow
1 1 0.3
2 1 0.2
3 1 0.2
4 1 0.2
6 1 0.1
5 2 0
```

**Reply by antoneri** [reply]
We've created a pull request to the maintainer of cdlib: https://github.com/GiulioRossetti/cdlib/pull/174

**Reply by GiulioRossetti** [reply]
Thanks for the pull request: the fix is now merged CDlib's master branch.

**Reply by JohannHM** [reply]
> I see, thanks for more details! You are using infomap through a third-party library `cdlib` that unfortunately [only loop over the edges](https://github.com/GiulioRossetti/cdlib/blob/master/cdlib/algorithms/crisp_partition.py#L720) in the graph when they add the network to Infomap. This is why disconnected nodes are missing in the output.
> 
> I didn't know about that library before, thanks for the information, I will make a pull request so they can fix that, and also suggest that they allow parameters to be sent to Infomap to support for example directed networks.
> 
> Until then, you can use Infomap directly, with the Python API documented [here](https://mapequation.github.io/infomap/python/). To include disconnected nodes you just have to call [add_node](https://mapequation.github.io/infomap/python/infomap.html#infomap.Infomap.add_node).
> 
> On the Infomap web app, disconnected nodes should be included in the output. How did you test it when you got the same result?

Hi  @danieledler Thanks a lot. Look, I tried to directly use your python implementation, but some methods seems not to work. May be for my python version Python 3.8? I don't know. So I gave the same network above to the infomap web and nodes [11, 17, 21] (the disconnected nodes) does not appear in any community when run it with flags (--clu --ftree).
Thank so much for open an issue in the cdlib.

**Reply by danieledler** [reply]
@JohannHM, if some methods don't seem to work in the Infomap python package, can you please open a Github issue for that where you describe the issue in a bit more detail? (I also use python 3.8, that shouldn't be any problem.)

Regarding the missing disconnected nodes, I used your code above and wrote a Pajek network with NetworkX (`nx.write_pajek`). For the `*vertices` section, I removed the meta data columns to only keep new and original node ids, so the original node ids (second column) will end up as node names in Infomap. Then I get this network:

```
*vertices 31
1 3
2 4
3 5
4 6
5 7
6 8
7 9
8 10
9 11
10 12
11 13
12 14
13 15
14 16
15 17
16 18
17 19
18 20
19 21
20 22
21 23
22 24
23 25
24 26
25 27
26 28
27 29
28 30
29 31
30 32
31 33
*edges
1 5 1.0
1 10 1.0
1 11 1.0
2 4 1.0
2 8 1.0
3 4 1.0
3 8 1.0
3 14 1.0
4 14 1.0
6 28 1.0
6 30 1.0
6 31 1.0
7 31 1.0
11 31 1.0
12 30 1.0
12 31 1.0
13 30 1.0
13 31 1.0
16 30 1.0
16 31 1.0
17 31 1.0
18 30 1.0
18 31 1.0
20 30 1.0
20 31 1.0
21 23 1.0
21 25 1.0
21 27 1.0
21 30 1.0
21 31 1.0
22 23 1.0
22 25 1.0
22 29 1.0
23 29 1.0
24 27 1.0
24 31 1.0
25 31 1.0
26 29 1.0
26 31 1.0
27 30 1.0
27 31 1.0
28 30 1.0
28 31 1.0
29 30 1.0
29 31 1.0
30 31 1.0
```

Running this network in [Infomap online](https://www.mapequation.org/infomap/) gives this tree:

```
# path flow name node_id
1:1 0.184783 "33" 31
1:2 0.119565 "32" 30
1:3 0.0543478 "23" 21
1:4 0.0543478 "31" 29
1:5 0.0434783 "29" 27
1:6 0.0326087 "8" 6
1:7 0.0326087 "24" 22
1:8 0.0326087 "25" 23
1:9 0.0326087 "27" 25
1:10 0.0326087 "30" 28
1:11 0.0217391 "14" 12
1:12 0.0217391 "15" 13
1:13 0.0217391 "18" 16
1:14 0.0217391 "20" 18
1:15 0.0217391 "22" 20
1:16 0.0217391 "26" 24
1:17 0.0217391 "28" 26
1:18 0.0108696 "9" 7
1:19 0.0108696 "19" 17
2:1 0.0326087 "5" 3
2:2 0.0326087 "6" 4
2:3 0.0217391 "4" 2
2:4 0.0217391 "10" 8
2:5 0.0217391 "16" 14
3:1 0.0326087 "3" 1
3:2 0.0217391 "13" 11
3:3 0.0108696 "7" 5
3:4 0.0108696 "12" 10
4:1 0 "11" 9
5:1 0 "17" 15
6:1 0 "21" 19
```

In the bottom, we have three singleton modules for the three disconnected nodes you mentioned.

So unfortunately I cannot reproduce the problem you get with disconnected nodes, maybe they got lost when you wrote the network to load in Infomap online? If you paste the network as @antoneri asked, or describe the steps to export it, I think we can sort the issue out!

**Reply by JohannHM** [reply]
Thanks @danieledler I tried to use the cdlib implementation because of the simplicity and because cdlib is based on your algorithm... I didn't want to write and read my networks in pajek each time because I have many networks and modifications to capture each time I run infomap. But let me see now how to pass the network to your algorithm without writing pajek file. When finding the way I'll be back to you guys if some question raises or if everithing is ok.

**Reply by danieledler** [reply]
Sure, I guess you'll find out, but after testing your code I replaced cdlib with infomap like below:

```py
im = infomap.Infomap()
im.add_nodes(G.nodes)
for u, v, data in G.edges(data=True):
    if 'weight' in data:
        im.add_link(u, v, data['weight'])
    else:
        im.add_link(u, v)

im.run()
recovering_nodes = [node for node, module in im.modules]

```

**Reply by JohannHM** [reply]
Everything is fine when working with in-built karate-club graph. When working with my own networks like that one in the  attachment, the following errors appears:
[G1_bal_rq3.pkl.tar.gz](https://github.com/mapequation/infomap/files/6609859/G1_bal_rq3.pkl.tar.gz)
```
TypeError: Wrong number or type of arguments for overloaded function 'InfomapWrapper_addNode'.
  Possible C/C++ prototypes are:
    infomap::InfomapWrapper::addNode(unsigned int)
    infomap::InfomapWrapper::addNode(unsigned int,std::string)
    infomap::InfomapWrapper::addNode(unsigned int,double)
    infomap::InfomapWrapper::addNode(unsigned int,std::string,double)
```
Is there any clue to cope this? I didn't understand this kind of error. Would you please introduce me some light about it?
Thanks in advance

**Reply by JohannHM** [reply]
> Everything is fine when working with in-built karate-club graph. When working with my own networks like that one in the attachment, the following errors appears:
> [G1_bal_rq3.pkl.tar.gz](https://github.com/mapequation/infomap/files/6609859/G1_bal_rq3.pkl.tar.gz)
> 
> ```
> TypeError: Wrong number or type of arguments for overloaded function 'InfomapWrapper_addNode'.
>   Possible C/C++ prototypes are:
>     infomap::InfomapWrapper::addNode(unsigned int)
>     infomap::InfomapWrapper::addNode(unsigned int,std::string)
>     infomap::InfomapWrapper::addNode(unsigned int,double)
>     infomap::InfomapWrapper::addNode(unsigned int,std::string,double)
> ```
> 
> Is there any clue to cope this? I didn't understand this kind of error. Would you please introduce me some light about it?
> Thanks in advance

Should it be with the fact that my node labels are not integers (like in karate-club), but strings?

**Reply by JohannHM** [reply]
mmm. I think something is working but I didn't understand why.
Look, my working graphs are composed by node labels labeled as strings of integer numbers, something like ['02', '368', '000', '056', etc]. When I change these line of your code:

`im.add_nodes(G.nodes)` by this:
```python
nodos = list(G.nodes)
losnodos = [int(nn) for nn in nodos]
```
To convert them into integer labels.
Likewise, when I modify the code to convert strings nodes to integer version:
```python
for u, v, data in G.edges(data=True):
    if 'weight' in data:        
        im.add_link(int(u), int(v), data['weight'])
    else:
        im.add_link(int(u), int(v))
```
The code runs without problem, but I'm unsure whether the problem is that your code only accept node labels as integer number but it does not accept node labels as strings. Is that true?

**Reply by antoneri** [reply]
Yes, you are correct that they need to be integers.

If you want to assign node labels as well, you can use the second argument:

```python
im = Infomap()
im.add_node(1, "label 1")
im.add_node(2, "label 2")
# or:
#nodes = [(1, "label 1"), (2, "label 2")]
#im.add_nodes(nodes)
im.add_link(1, 2)
```

So in your case, you can map the labels to integers:
```python
nodes = list(G.nodes)
node_map = {label: node for node, label in enumerate(nodes)}

for label, node in node_map.items():
    im.add_node(node, label)

for u, v, data in G.edges(data=True):
    node_u, node_v = node_map[u], node_map[v]
    if 'weight' in data:        
        im.add_link(node_u, node_v, data['weight'])
    else:
        im.add_link(node_u, node_v)
```

Hope this helps!

**Reply by JohannHM** [reply]
Hello again.
It helped a lot. Thank so much @antoneri and @danieledler for your helping hand.Very glad of discover the big usefulness of your code.
Johann

**Reply by danieledler** [reply]
Hi @JohannHM,

Glad to hear. By the way, after our last reply above on relabelling node ids to integers, we added a helper method to [add a NetworkX graph to Infomap](https://mapequation.github.io/infomap/python/infomap.html#infomap.Infomap.add_networkx_graph) that works for both integer ids and string ids.

If the node ids are strings, they will be added as node names, but the `add_networkx_graph(G)` will also return a dictionary that maps the new node ids to the string labels:

```py
G = nx.Graph([("a", "b"), ("a", "c")])
im = Infomap()
mapping = im.add_networkx_graph(G)
mapping
# {0: 'a', 1: 'b', 2: 'c'}

im.run()
for node in im.nodes:
    print(node.node_id, node.module_id, node.flow, im.get_name(node.node_id))
# Output:
# 0 1 0.5 a
# 1 1 0.25 b
# 2 1 0.25 c
# 
```

Should be useful in your case, so please try!

**Reply by JohannHM** [reply]
Hi again guys!
Off-course, after working with the `add_networkx_graph(g)` method, I got the same results. However, this method makes the life easier for those who are not python experts like me.
Thank you so much for this implementation guys. Hope to get more interesting results and honored to cite you all.
btw: for sure I'll contact you again for more foo, questions.
